bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,365 @@
+const PORT_ARG: &str = "port";
+const FILE_ARG: &str = "file";
+
+const PUT_SC: &str = "put";
+const GET_SC: &str = "get";
+const EXISTS_SC: &str = "exists";
+
+#[derive(Debug)]
+enum ParseError {
+    FilePattern,
+    KeyPattern,
+}
+
+#[derive(Debug)]
+struct AppArgs<'a> {
+    subcommand: &'a str,
+    file_pattern: Option<FilePattern<'a>>,
+    key_pattern: Option<KeyPattern>,
+    keysize: usize,
+    host: String,
+    port: u16,
+}
+
+impl<'a> AppArgs<'a> {
+    fn from_matches(matches: &'a ArgMatches<'a>) -> Self {
+        let (subcommand, sub_matches) = matches.subcommand();
+        let sub_matches = sub_matches.unwrap();
+        let file_pattern = match sub_matches.value_of(FILE_ARG) {
+            None => None,
+            Some(p) => match Self::parse_file_pattern(p) {
+                Ok(val) => Some(val),
+                Err(e) => panic!("{:?}", e), // should send an error to the caller
+            },
+        };
+        let key_pattern = match sub_matches.value_of(KEY_ARG) {
+            None => None,
+            Some(p) => match Self::parse_key_pattern(p) {
+                Ok(val) => Some(val),
+                Err(e) => panic!("{:?}", e), // again
+            },
+        };
+
+        Self {
+            subcommand,
+            file_pattern,
+            key_pattern,
+            keysize: sub_matches.value_or_default(KEY_SIZE_ARG),
+            host: sub_matches.value_or_default(HOST_ARG),
+            port: sub_matches.value_or_default(PORT_ARG),
+        }
+    }
+    fn parse_file_pattern(file_pattern: &str) -> Result<FilePattern<'_>, ParseError> {
+        let path = PathBuf::from(file_pattern);
+        let name = path.file_name().unwrap().to_str().unwrap();
+
+        let re = Regex::new("\\{key\\}").unwrap();
+        let two_or_more = Regex::new(".*\\{key\\}.*\\{key\\}.*").unwrap();
+
+        if re.is_match(name) {
+            if two_or_more.is_match(name) {
+                Err(ParseError::FilePattern)
+            } else {
+                let path = PathBuf::from(file_pattern);
+
+                let dir = match path.parent().unwrap() {
+                    p if p == Path::new("") => Path::new("."),
+                    p => p,
+                };
+                Ok(FilePattern::WithRE(
+                    RePattern::new(file_pattern),
+                    dir.to_owned(),
+                ))
+            }
+        } else {
+            Ok(FilePattern::WithoutRE(file_pattern))
+        }
+    }
+    fn parse_key_pattern(key_arg_cont: &str) -> Result<KeyPattern, ParseError> {
+        if let Ok(key) = key_arg_cont.parse() {
+            return Ok(KeyPattern::Single(key));
+        }
+        if key_arg_cont.contains("-") {
+            let v: Vec<&str> = key_arg_cont.split("-").collect();
+            if v.len() != 2 || v[0] >= v[1] {
+                return Err(ParseError::KeyPattern);
+            }
+            let v: Vec<u64> = v.into_iter().map(|n| n.parse().unwrap()).collect();
```

Comment:
At this point you've already know that there is 2 items in `v` vector, thus you don't need an iterator and call to `collect`, you can directly parse values by indexes (`v[0].parse().unwrap()..=v[1].parse().unwrap()`)

Reply:
Fix: [05d1376](https://github.com/qoollo/bob/pull/562/commits/05d1376ca40d503396e60d53b469dbf8bf7d4878)